src/nnet3/nnet-simple-component.cc

```diff
@@ -119,16 +136,35 @@ void DropoutComponent::Propagate(const ComponentPrecomputedIndexes *indexes,
 
   BaseFloat dropout = dropout_proportion_;
   KALDI_ASSERT(dropout >= 0.0 && dropout <= 1.0);
-
-  // This const_cast is only safe assuming you don't attempt
-  // to use multi-threaded code with the GPU.
-  const_cast<CuRand<BaseFloat>&>(random_generator_).RandUniform(out);
-
-  out->Add(-dropout); // now, a proportion "dropout" will be <0.0
-  out->ApplyHeaviside(); // apply the function (x>0?1:0).  Now, a proportion "dropout" will
-                         // be zero and (1 - dropout) will be 1.0.
-
-  out->MulElements(in);
+  if(!dropout_per_frame_)
+  {
+    // This const_cast is only safe assuming you don't attempt
+    // to use multi-threaded code with the GPU.
+    const_cast<CuRand<BaseFloat>&>(random_generator_).RandUniform(out);
+
+    out->Add(-dropout); // now, a proportion "dropout" will be <0.0
+    out->ApplyHeaviside(); // apply the function (x>0?1:0).  Now, a proportion "dropout" will
+                          // be zero and (1 - dropout) will be 1.0.
+    out->MulElements(in);
+  } else {
+    // randomize the dropout matrix by row,
+    // i.e. [[1,1,1,1],[0,0,0,0],[0,0,0,0],[1,1,1,1],[0,0,0,0]]
+    // This const_cast is only safe assuming you don't attempt
+    // to use multi-threaded code with the GPU.
+    CuSubMatrix<BaseFloat> out_col_submatrix = out->ColRange(0, 1);
+    const_cast<CuRand<BaseFloat>&>(random_generator_).RandUniform(&out_col_submatrix); 
```

Comment:
@danpovey  `CuMatrix tmp(1, out->NumCols(), kUndefined); ` ? or We should use `CuMatrix tmp(out->NumRow(), 1, kUndefined);`  We want to copy the same col to the whole matrix(so matrix will be the same in row dim), the tmp should be row_num*1, not 1*col_num. Is this right?